### PR TITLE
8571: raise TypeError for expressions with BooleanAtom

### LINKED ISSUE #8441: Multiplying a boolean by a float
Sympy does not support multiplying integers and booleans:
```python
S(3)*S(True)
```
raises
```pytb
Traceback (most recent call last):
  File "<ipython-input-3-00b7489808d9>", line 1, in <module>
    S(3)*S(True)
  File "sympy/core/numbers.py", line 1671, in __mul__
    return Rational.__mul__(self, other)
  File "sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "sympy/core/numbers.py", line 1227, in __mul__
    return Number.__mul__(self, other)
  File "sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "sympy/core/numbers.py", line 361, in __mul__
    return AtomicExpr.__mul__(self, other)
  File "sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "sympy/core/decorators.py", line 118, in binary_op_wrapper
    return func(self, other)
  File "sympy/core/expr.py", line 141, in __mul__
    return Mul(self, other)
  File "sympy/core/cache.py", line 91, in wrapper
    retval = cfunc(*args, **kwargs)
  File "sympy/core/compatibility.py", line 872, in wrapper
    result = user_function(*args, **kwds)
  File "sympy/core/operations.py", line 41, in __new__
    c_part, nc_part, order_symbols = cls.flatten(args)
  File "sympy/core/mul.py", line 184, in flatten
    r, b = b.as_coeff_Mul()
AttributeError: 'BooleanTrue' object has no attribute 'as_coeff_Mul'
```

However, multiplying floats by booleans does work for some reason:
```python
In [4]: S(3.)*S(True)
Out[4]: 3.0*True
```

This may lead to future exceptions: `_.evalf()` raises
```pytb
Traceback (most recent call last):
  File "<ipython-input-5-11d3aed22e4e>", line 1, in <module>
    _.evalf()
  File "sympy/core/evalf.py", line 1320, in evalf
    v = self._eval_evalf(prec)
  File "sympy/core/mul.py", line 614, in _eval_evalf
    rv = AssocOp._eval_evalf(self, prec)
  File "sympy/core/operations.py", line 332, in _eval_evalf
    newa = a._eval_evalf(prec)
AttributeError: 'BooleanTrue' object has no attribute '_eval_evalf'
```

Is this on purpose? Or is it a bug?

### LINKED ISSUE #8571: printing: Add containing a Boolean fails
I offer no particular opinion on whether they make sense but neither of these operations seem to be forbidden:
````
a = x*S.true
b = x + S.true
````
But only `a` can be printed (and gives `x*True`) whereas `print(b)` and `pprint(b)` give:

````
AttributeError: 'BooleanTrue' object has no attribute 'as_coeff_Mul'
````

### PR DESCRIPTION
**Pull request by smichr**
closes #8571
closes #8441
closes #10569

The `_op_priority` of BooleanAtom is set to 1 greater than Expr so
its `__r*__` methods get called first and allow a consistent
TypeError to be raised for expressions involving S.true or S.false.

### CODE REVIEW

File: sympy/logic/boolalg.py [hunk 1]
```diff
@@ -84,10 +84,31 @@ class BooleanAtom(Boolean):
...
     @property
     def canonical(self):
         return self
 
+    def _noop(self, other=None):
+        raise TypeError('BooleanAtom not allowed in this context.')
+
+    __add__ = _noop
+    __radd__ = _noop
+    __neg__ = _noop
```
Comment (mvnnn) [inline sympy/logic/boolalg.py]: we not need ```__neg__``` because it's already create a good```TypeError```.

File: sympy/logic/boolalg.py [hunk 2]
```diff
@@ -84,10 +84,31 @@ class BooleanAtom(Boolean):
...
+    __rmul__ = _noop
+    __pow__ = _noop
+    __rpow__ = _noop
+    __rdiv__ = _noop
+    __truediv__ = _noop
+    __div__ = _noop
+    __rtruediv__ = _noop
+    __mod__ = _noop
+    _eval_power = _noop
+
```
Comment (mvnnn) [inline sympy/logic/boolalg.py]: add ```__invert__``` here.

Comment (smichr) [inline sympy/logic/boolalg.py]: added `__rmod__` and deleted `__neg__`

**Comment by jksuom** [comment]
Looks good. +1

**Comment by mvnnn** [comment]
+1

**Pull request merged**